"The following types implement trait..." suggets crate private items, if inside macro.

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

There are two crates here, `library` and `consumer` (which depends on `library`).

library:
```rust
#![feature(decl_macro)]

pub macro global_meow($item:expr) {
    const _: () = { 
        fn foo() {
            Meow::meow(&$item);
        }
    };
}

#[macro_export]
macro_rules! global_meow_rules {
    ($item:expr) => {
        const _: () = { 
            fn foo() {
                $crate::Meow::meow(&$item);
            }
        };
    }
}

pub trait Meow {
    fn meow(&self) {}
}

pub struct GlobalMeow;

impl Meow for GlobalMeow { }

pub(crate) struct PrivateMeow;

impl Meow for PrivateMeow { }
```

consumer:
```rust
static NOT_MEOW: usize = 0;

library::global_meow!(NOT_MEOW);

library::global_meow_rules!(NOT_MEOW);

fn main() {}
```

The current output is (when running `cargo check` on `consumer`):

```rust
# cargo check
    Checking consumer v0.1.0 (/tmp/scratcho2FpPeJDq/consumer)
error[E0277]: the trait bound `usize: Meow` is not satisfied
 --> src/main.rs:2:1
  |
2 | library::global_meow!(NOT_MEOW);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Meow` is not implemented for `usize`
  |
  = help: the following other types implement trait `Meow`:
            GlobalMeow
            library::PrivateMeow
  = note: this error originates in the macro `library::global_meow` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `usize: Meow` is not satisfied
 --> src/main.rs:4:1
  |
4 | library::global_meow_rules!(NOT_MEOW);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Meow` is not implemented for `usize`
  |
  = help: the following other types implement trait `Meow`:
            GlobalMeow
            library::PrivateMeow
  = note: this error originates in the macro `library::global_meow_rules` (in Nightly builds, run with -Z macro-backtrace for more info)
```

<!-- The following is not always necessary. -->
Ideally the output should look like:

```
# cargo check
    Checking consumer v0.1.0 (/tmp/scratcho2FpPeJDq/consumer)
error[E0277]: the trait bound `usize: Meow` is not satisfied
 --> src/main.rs:2:1
  |
2 | library::global_meow!(NOT_MEOW);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Meow` is not implemented for `usize`
  |
  = help: the following other types implement trait `Meow`:
            GlobalMeow
  = note: this error originates in the macro `library::global_meow` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `usize: Meow` is not satisfied
 --> src/main.rs:4:1
  |
4 | library::global_meow_rules!(NOT_MEOW);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Meow` is not implemented for `usize`
  |
  = help: the following other types implement trait `Meow`:
            GlobalMeow
  = note: this error originates in the macro `library::global_meow_rules` (in Nightly builds, run with -Z macro-backtrace for more info)
```

We should not be suggesting private types to users, since macro hygiene *does* mean they can't write `library::PrivateMeow`

This is on the latest nightly, `rustc 1.64.0-nightly (06754d885 2022-07-08)`

This was ran into in https://github.com/rust-lang/rust/pull/99074

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
@@ -673,6 +673,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                             if !self.report_similar_impl_candidates(
                                 impl_candidates,
                                 trait_ref,
+                                obligation.cause.body_id,
                                 &mut err,
                             ) {
                                 // This is *almost* equivalent to
@@ -707,6 +708,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                                     self.report_similar_impl_candidates(
                                         impl_candidates,
                                         trait_ref,
+                                        obligation.cause.body_id,
                                         &mut err,
                                     );
                                 }
@@ -1353,6 +1355,7 @@ trait InferCtxtPrivExt<'hir, 'tcx> {
         &self,
         impl_candidates: Vec<ImplCandidate<'tcx>>,
         trait_ref: ty::PolyTraitRef<'tcx>,
+        body_id: hir::HirId,
         err: &mut Diagnostic,
     ) -> bool;
 
@@ -1735,6 +1738,7 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
         &self,
         impl_candidates: Vec<ImplCandidate<'tcx>>,
         trait_ref: ty::PolyTraitRef<'tcx>,
+        body_id: hir::HirId,
         err: &mut Diagnostic,
     ) -> bool {
         let report = |mut candidates: Vec<TraitRef<'tcx>>, err: &mut Diagnostic| {
@@ -1816,11 +1820,9 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
                         // FIXME(compiler-errors): This could be generalized, both to
                         // be more granular, and probably look past other `#[fundamental]`
                         // types, too.
-                        match self.tcx.visibility(def.did()) {
-                            ty::Visibility::Public => true,
-                            ty::Visibility::Restricted(def_id) => def_id.is_local(),
-                            ty::Visibility::Invisible => false,
-                        }
+                        self.tcx
+                            .visibility(def.did())
+                            .is_accessible_from(body_id.owner.to_def_id(), self.tcx)
                     } else {
                         true
                     }

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
@@ -1805,8 +1805,26 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
                         || self.tcx.is_builtin_derive(def_id)
                 })
                 .filter_map(|def_id| self.tcx.impl_trait_ref(def_id))
-                // Avoid mentioning type parameters.
-                .filter(|trait_ref| !matches!(trait_ref.self_ty().kind(), ty::Param(_)))
+                .filter(|trait_ref| {
+                    let self_ty = trait_ref.self_ty();
+                    // Avoid mentioning type parameters.
+                    if let ty::Param(_) = self_ty.kind() {
+                        false
+                    }
+                    // Avoid mentioning types that are private to another crate
+                    else if let ty::Adt(def, _) = self_ty.peel_refs().kind() {
+                        // FIXME(compiler-errors): This could be generalized, both to
+                        // be more granular, and probably look past other `#[fundamental]`
+                        // types, too.
+                        match self.tcx.visibility(def.did()) {
+                            ty::Visibility::Public => true,
+                            ty::Visibility::Restricted(def_id) => def_id.is_local(),
+                            ty::Visibility::Invisible => false,
+                        }
+                    } else {
+                        true
+                    }
+                })
                 .collect();
             return report(normalized_impl_candidates, err);
         }

--- a/src/test/ui/suggestions/auxiliary/meow.rs
+++ b/src/test/ui/suggestions/auxiliary/meow.rs
@@ -1,0 +1,11 @@
+pub trait Meow {
+    fn meow(&self) {}
+}
+
+pub struct GlobalMeow;
+
+impl Meow for GlobalMeow {}
+
+pub(crate) struct PrivateMeow;
+
+impl Meow for PrivateMeow {}

--- a/src/test/ui/suggestions/issue-99080.rs
+++ b/src/test/ui/suggestions/issue-99080.rs
@@ -1,0 +1,16 @@
+// aux-build:meow.rs
+
+extern crate meow;
+
+use meow::Meow;
+
+fn needs_meow<T: Meow>(t: T) {}
+
+fn main() {
+    needs_meow(1usize);
+    //~^ ERROR the trait bound `usize: Meow` is not satisfied
+}
+
+struct LocalMeow;
+
+impl Meow for LocalMeow {}

--- a/src/test/ui/suggestions/issue-99080.stderr
+++ b/src/test/ui/suggestions/issue-99080.stderr
@@ -1,0 +1,20 @@
+error[E0277]: the trait bound `usize: Meow` is not satisfied
+  --> $DIR/issue-99080.rs:10:16
+   |
+LL |     needs_meow(1usize);
+   |     ---------- ^^^^^^ the trait `Meow` is not implemented for `usize`
+   |     |
+   |     required by a bound introduced by this call
+   |
+   = help: the following other types implement trait `Meow`:
+             GlobalMeow
+             LocalMeow
+note: required by a bound in `needs_meow`
+  --> $DIR/issue-99080.rs:7:18
+   |
+LL | fn needs_meow<T: Meow>(t: T) {}
+   |                  ^^^^ required by this bound in `needs_meow`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0277`.

--- a/src/test/ui/traits/bound/assoc-fn-bound-root-obligation.stderr
+++ b/src/test/ui/traits/bound/assoc-fn-bound-root-obligation.stderr
@@ -15,7 +15,6 @@ LL |     s.strip_suffix(b'\n').unwrap_or(s)
              &'c &'b str
              [char; N]
              char
-             pattern::MultiCharEqPattern<C>
    = note: required because of the requirements on the impl of `Pattern<'_>` for `u8`
 
 error: aborting due to previous error

